Feat power button

sparchatus opened the pull request:
Add support for ACPI fixed power buttons. This is based on https://github.com/KernelTestFramework/ktf/pull/350

I wanted to add the functionality to change the power button handler because it would be useful for triggering experiments for me. This opened up a race condition between the interrupt and the function changing the handler. The interrupt may not wait for a spinlock so I added a try_lock function but I am not very convinced by this approach.

sparchatus committed:
drivers: ACPI fixed power button driver

Driver for ACPI fixed power buttons. It allows
overwriting the default handler so the button can
be used for custom purposes. One such purpose might
be to trigger an experiment to run.

Signed-off-by: Sandro Rüegge <rueegges@ethz.ch>

sparchatus committed:
cmdline: add option to enable power button handling

Signed-off-by: Sandro Rüegge <rueegges@ethz.ch>

wipawel commented on include/spinlock.h:
```diff
@@ -37,9 +37,14 @@ typedef volatile unsigned int spinlock_t;
 /* Static declarations */
 
+static inline bool spin_try_lock(spinlock_t *lock) {
+    ASSERT(lock);
```
nit: since this runs in a loop, I wouldn't place this assert here. Probably better to leave it to a caller.

sparchatus replied:
removed it

wipawel commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    AcpiClearEvent(ACPI_EVENT_POWER_BUTTON);
+
+    if (spin_try_lock(&pb_handler_lock)) {
+        if (pb_handler)
```
probably to better to take this check out of the CS and bail early without touching locks if no handler defined

sparchatus replied:
I believe that could lead to a race condition if we first check non-null but then the handler is set to NULL by a different core before we get the spinlock. Maybe the easiest solution is what minipli-oss proposed if we are willing to limit handler changes to the bsp :thinking:

wipawel replied:
Yes, you are right. I was operating with the assumption that this is touched by BSP only. Essentially, the `pb_set_handler` usage should be careful, ideally before the SCI is configured/unmasked (but that's problematic) or could sync all CPUs via IPI mechanism (a bit of an overkill, and unimplemented). Maybe limiting the use to BSP (via proper function comments) would indeed suffice.

wipawel commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    return ACPI_INTERRUPT_HANDLED;
+}
+
+static void default_handler(void *) {
```
nit: missing function parameter name, add `notused` or something like this

wipawel commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    ACPI_TABLE_FADT *fadt = acpi_find_table(ACPI_SIG_FADT);
+
+    if (!(fadt->Flags & ACPI_FADT_POWER_BUTTON)) {
+        printk("PWRB: fixed button handler\n");
```
nit: maybe it's just me, but I find this message a bit confusing

sparchatus replied:
oups, I see the problem :smile:

wipawel commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    spin_unlock(&pb_handler_lock);
+}
+
+bool init_power_button(const cpu_t *cpu) {
```
nit: is the `cpu` param used?

wipawel commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    if (!(fadt->Flags & ACPI_FADT_POWER_BUTTON)) {
+        printk("PWRB: fixed button handler\n");
+
+        if (ACPI_FAILURE(AcpiClearEvent(ACPI_EVENT_POWER_BUTTON))) {
```
There should be some `KTF_ACPICA` preprocessor checks so this code compiles without ACPICA. OTOH... the non-ACPICA variant is there just to help understand why KTF does not boot on some machines... so the checks can be kept very simple here.

sparchatus replied:
I added the preprocessor checks and changed the power button to default to disabled.

minipli-oss commented on drivers/power_button.c:
```diff
@@ -0,0 +1,57 @@
+    spin_lock(&pb_handler_lock);
+    pb_handler = handler;
+    pb_context = context;
+    spin_unlock(&pb_handler_lock);
```
I guess, SCI is just a regular interrupt? If so, wrapping above assignment in `CLI ... STI` should be sufficient to make the assignment atomic wrt. its use in button_handler(), vanishing the need for the lock altogether. E.g., something like this:

```c
static UINT32 button_handler(void *Context) {
    AcpiClearEvent(ACPI_EVENT_POWER_BUTTON);

    if (pb_handler)
        pb_handler(pb_context);

    return ACPI_INTERRUPT_HANDLED;
}

void pb_set_handler(pb_handler_t handler, void *context) {
    unsigned long flags = interrupts_disable_save();
    pb_handler = handler;
    pb_context = context;
    interrupts_restore(flags);
}
```

Not having a try-lock in button_handler() also prevents loosing events for the case where pb_set_handler() got interrupted by an SCI while holding the lock. With above code, that IRQ would be deferred until interrupts are enabled again.

Or are you concerned about one CPU calling pb_set_handler() while another one is serving the SCI?

sparchatus replied:
Good point, this would make it much simpler!  We direct the SCI handling to the bsp cpu so if we also restrict the pb_set_handler function to bsp CPU, then this should be fine. :thinking:

sparchatus replied:
Thanks again. I have applied your suggestion.

sparchatus committed:
percpu,drivers: Add cpu self reference to percpu

Adding a self reference to the percpu allows
checking what CPU some code is currently
running on. We use this in pb_set_handler
to ensure it runs on the bsp.

minipli-oss commented on drivers/power_button.c:
```diff
@@ -0,0 +1,60 @@
+
+static void default_handler(void *);
+
+spinlock_t pb_handler_lock = SPINLOCK_INIT;
```
`pb_handler_lock` isn't used any more, can remove its definition as well.

sparchatus replied:
oups, removed it

minipli-oss commented on include/percpu.h:
```diff
@@ -60,6 +61,7 @@ struct percpu {
     unsigned long usermode_private;
     volatile unsigned long apic_ticks;
     bool apic_timer_enabled;
+    struct cpu *cpu;
```
I guess, using the proper typedef'd type `cpu_t` would cause a cyclic header include dependency?

sparchatus replied:
Yes. I could add the `cpu_t` typedef to percpu.h, though...

minipli-oss reviewed (changes requested):
Otherwise LGTM!

minipli-oss commented on drivers/power_button.c:
```diff
@@ -0,0 +1,59 @@
+#include <acpi_ktf.h>
+#include <drivers/power_button.h>
+#include <spinlock.h>
```
This include shouldn't be needed any more

sparchatus replied:
thanks, that was sloppy. I now also removed an obsolete include from power_button.h, fixed compilation issues when not using acpica and power button init error will warn and return `false` instead of panicking as it is not an essential OS component.

minipli-oss reviewed (approved):
LGTM!

Thanks, @sparchatus!
